router next documentation (#6045)
apollographql/router: docs/source/executing-operations/native-query-planner.mdx

---
title: Native Query Planner
subtitle: Run the Rust-native query planner in GraphOS Router
minVersion: 1.49.0
---

<PreviewFeature />

Learn to run the GraphOS Router with the Rust-native query planner and improve your query planning performance and scalability.

## Background about query planner implementations

> In v1.49.0 the router introduced a [query planner](/graphos/routing/about-router#query-planning) implemented natively in Rust. This native query planner improves the overall performance and resource utilization of query planning. It exists alongside the legacy Javascript implementation. and it will eventually replace the legacy implementation.

Reviewer: ```suggestion
In v1.49.0 the router introduced a [query planner](/graphos/routing/about-router#query-planning) implemented natively in Rust. This native query planner improves the overall performance and resource utilization of query planning. It exists alongside the legacy Javascript implementation, and it will eventually replace the legacy implementation.
```

> As part of the effort to ensure correctness and stability of the new query planner, starting in v1.53.0 the router can enable both the new and legacy planners and run them in parallel to compare their results. After their comparison, the router discards the native query planner's results and uses only the legacy planner to execute requests.

Reviewer: So this happens by _default_ since 1.53.0. We are not switching the default of running both side by side for this release.
```suggestion
As part of the effort to ensure correctness and stability of the new query planner, starting in v1.53.0 the router enables both the new and legacy planners and runs them in parallel to compare their results by default. After their comparison, the router discards the native query planner's results and uses only the legacy planner to execute requests.
```

> The native query planner uses a single thread in the cold path of the router. It has a bounded queue of ten queries. If the queue is full, the router simply does not run the comparison to avoid excessive resource consumption.

Reviewer: I am afraid this is not making it clear that this only happens in the comparison - `both` - mode 🤔 do you think we should change the title of this section to mention that this is only in comparison (which does run by default at the moment)

## Configuring query planning

You can configure the `experimental_query_planner_mode` option in your `router.yaml` to set the query planner to run.

The `experimental_query_planner_mode` option has the following supported modes:

- `new`- enables only the new Rust-native query planner
- `legacy` - enables only the legacy JavaScript query planner 
- `both_best_effort` - enables both new and legacy query planners for comparison. The legacy query planner is used for execution. 

<MinVersion version="2.0.0">

## Optimize native query planner

> </MinVersion>

Reviewer 2: @lrlna is this correct that 2.0.0 is the minimum router version that supports all-native query planning?

Reviewer 3: No, will be available from 1.56.0

> To run the native query planner with the best performance and resource utilization, configure your router with the following options:

Reviewer: This config not only enable native query planning, but also _does not_ startup v8, the javascript runtime, which is what we have to run in order to run the legacy query planner.

v8 eats through a lot of CPU and memory in a router process when it's running, and now that we have confidence in our native implementation, that we want users to disable it when running the router. This is the config that gets them the native query planner, and does not start v8.

Reviewer 2: edited

```yaml title="router.yaml"
supergraph:
  generate_query_fragments: true

> 

Reviewer: I've been thinking about this some more, and now I think this should be a separate point from the call to action of enabling the rust planner and disabling v8. Perhaps we can say something like:
> Additionally, to enable faster throughput by reducing the size of queries sent to subgraphs, and more optimal native query planning, enable query fragment generation:
>```yaml title="router.yaml"
> supergraph:
>   generate_query_fragments: true
>```
What do you think?

Reviewer 3: Agree - changing too many things at once might feel scary. Let's have clear outcomes 
* Test new QP
* Disable v8

experimental_query_planner_mode: new
experimental_introspection_mode: new
```

- `experimental_query_planner_mode: new` enables only the Rust-native query planner.
- `generate_query_fragments: true` enables faster throughput by reducing the size of queries sent to subgraphs.
- `experimental_introspection_mode: new` enables Rust-native schema introspection.

## Metrics for native query planner

TBD

## Limitations of native query planner

> TBD

Reviewer 2: @lrlna is there any telemetry, limitations, are other topics that need to be called out on this page?

---
title: Native Query Planner
subtitle: Run the Rust-native query planner in GraphOS Router
minVersion: 1.49.0
---

<PreviewFeature />

Learn to run the GraphOS Router with the Rust-native query planner and improve your query planning performance and scalability.

## Background about query planner implementations

> In v1.49.0 the router introduced a [query planner](/graphos/routing/about-router#query-planning) implemented natively in Rust. This native query planner improves the overall performance and resource utilization of query planning. It exists alongside the legacy Javascript implementation that uses the V8 JavaScript engine, and it will eventually replace the legacy implementation.

Reviewer: ```suggestion
In v1.49.0 the router introduced a [query planner](/graphos/routing/about-router#query-planning) implemented natively in Rust. This native query planner improves the overall performance and resource utilization of query planning. It exists alongside the legacy JavaScript implementation that uses the V8 JavaScript engine, and it will eventually replace the legacy implementation.
```

### Comparing query planner implementations

> As part of the effort to ensure correctness and stability of the new query planner, starting in v1.53.0 the router can enable both the new and legacy planners and run them in parallel to compare their results by default. After their comparison, the router discards the native query planner's results and uses only the legacy planner to execute requests. The native query planner uses a single thread in the cold path of the router. It has a bounded queue of ten queries. If the queue is full, the router simply does not run the comparison to avoid excessive resource consumption.

Reviewer: It runs `both_best_effort` by default, rather than it being a possibility to run both by default. Do you think this would be more clear?

```suggestion
As part of the effort to ensure correctness and stability of the new query planner, starting in v1.53.0 the router enables both the new and legacy planners and runs them in parallel to compare their results by default. After their comparison, the router discards the native query planner's results and uses only the legacy planner to execute requests. The native query planner uses a single thread in the cold path of the router. It has a bounded queue of ten queries. If the queue is full, the router simply does not run the comparison to avoid excessive resource consumption.
```

## Configuring query planning

You can configure the `experimental_query_planner_mode` option in your `router.yaml` to set the query planner to run.

The `experimental_query_planner_mode` option has the following supported modes:

- `new`- enables only the new Rust-native query planner
- `legacy` - enables only the legacy JavaScript query planner 
- `both_best_effort` - enables both new and legacy query planners for comparison. The legacy query planner is used for execution. 

<MinVersion version="1.56.0">

## Optimize native query planner

</MinVersion>

To run the native query planner with the best performance and resource utilization, configure your router with the following options:

```yaml title="router.yaml"
experimental_query_planner_mode: new
experimental_introspection_mode: new
```

Setting `experimental_query_planner_mode: new` and `experimental_introspection_mode: new` not only enables native query planning and schema introspection, it also disables the V8 JavaScript runtime used by the legacy query planner. Disabling V8 frees up CPU and memory and improves native query planning performance.

> Additionally, to enable more optimal native query planning and faster throughput by reducing the size of queries sent to subgraphs, you can enable query fragment generation with the following option:

Reviewer: @chandrikas do you think should we mention here that this option will be the only option for handling fragments in the future?

also debating on whether we should like to our docs on [fragment reuse and generation](https://www.apollographql.com/docs/router/configuration/overview/#fragment-reuse-and-generation) and whether or not we should also update those docs under `generate_query_fragments` saying this will be the only way to deal with fragments in the future? 🤔 🤔 🤔

Reviewer 2: updated
